src/app/app.ts

```diff
@@ -1,71 +1,132 @@
 import {Post} from 'mattermost-redux/types/posts';
-import {AppCall, AppContext} from 'mattermost-redux/types/apps';
+import {AppCall, AppContext, AppCallValues, AppCallResponse} from 'mattermost-redux/types/apps';
 
-import {Env, tryPromiseWithMessage, errorWithMessage} from '../utils';
+import {SubscriptionFields} from '../utils/constants';
+
+import {
+    newErrorCallResponseWithFieldErrors,
+    newOKCallResponse,
+    newOKCallResponseWithMarkdown,
+    newErrorCallResponseWithMessage,
+    FieldValidationErrors} from '../utils/call_responses';
+import {Env, tryPromiseWithMessage} from '../utils';
 
 import {newMMClient, newZDClient} from '../clients';
 
-import {configStore, oauthStore} from '../store';
+import {configStore} from '../store';
 
-import {newTicketFromForm, FieldValidationErrors} from './model';
+import {newTicketFromForm} from './ticketFromForm';
+import {newTriggerFromForm} from './triggerFromForm';
 
-class App {
-    createTicketFromPost = async (call: AppCall): Promise<FieldValidationErrors> => {
-        // get active mattermost user ID
-        const mmUserID = call.context.acting_user_id || '';
-        const zdToken = oauthStore.getToken(mmUserID);
-        if (!zdToken) {
-            throw new Error('Failed to get user access_token');
-        }
+export interface IApp {
+    createTicketFromPost(): Promise<AppCallResponse>;
+    createZDSubscription(): Promise<AppCallResponse>;
+}
 
+class App implements IApp {
+    context: AppContext
+    values: AppCallValues
+
+    constructor(call: AppCall) {
+        this.context = call.context;
+        this.values = call.values;
+    }
+    createTicketFromPost = async (): Promise<AppCallResponse> => {
         // get zendesk client for user
-        const zdClient = newZDClient(zdToken);
+        const zdClient = newZDClient(this.context);
 
         // create the ticket object from the form response
-        const [zdTicket, errors] = newTicketFromForm(call.values);
+        const [zdTicketPayload, fieldErrors] = newTicketFromForm(this.values);
 
         // respond with errors
-        if (Object.keys(errors).length !== 0) {
-            return errors;
+        if (this.hasFieldErrors(fieldErrors)) {
+            return newErrorCallResponseWithFieldErrors(fieldErrors);
         }
 
         // create the ticket in Zendesk
-        const ticket = await tryPromiseWithMessage(zdClient.tickets.create(zdTicket), 'Failed to create Zendesk ticket');
+        const createReq = zdClient.tickets.create(zdTicketPayload);
+        const zdTicket = await tryPromiseWithMessage(createReq, 'Failed to create Zendesk ticket');
```

Comment:
What do you think is the correct course of action if this happens? As is, the `tryPromiseWithMessage` function will throw an error and the message will be shown in the modal as an error.
![image](https://user-images.githubusercontent.com/7575921/108091510-65eabd80-7041-11eb-9134-726b95201ba7.png)

Reply:
Maybe it was not clear to me that this throws an error and that it is catch somewhere else. If that is the behaviour, I am fine with it, but the code does not feel straight forward.

Reply:
also added to the error handling Jira ticket. I was trying to have a helper function that would to some error handling instead of having try/catch blocks all over the place. If this isn't clear, I'll adjust it and push for feedback in the error handling PR.